.pre-commit-config.yaml

```diff
@@ -26,10 +26,21 @@ repos:
                 files: python/.*\.(py|pyx|pxd)$
                 types: [file]
       - repo: https://github.com/pre-commit/mirrors-mypy
-        rev: 'v0.782'
+        rev: 'v0.971'
         hooks:
               - id: mypy
-                args: ["--config-file=setup.cfg", "python/cudf/cudf", "python/dask_cudf/dask_cudf", "python/custreamz/custreamz", "python/cudf_kafka/cudf_kafka"]
+                additional_dependencies: [types-cachetools==5.2.1]
+                args: ["--config-file=setup.cfg",
+                       "--exclude", "python/cudf/benchmarks",
```

Comment:
Can these be specified in the `setup.cfg` instead of the pre-commit flags? https://mypy.readthedocs.io/en/stable/config_file.html#confval-exclude

That way it will be more consistent if mypy is run outside of pre-commit.

Reply:
Done.

Reply:
Note that if you run mypy yourself you get errors that the mypy in pre-commit doesn't see (because your environment has numpy installed so mypy has more type annotations in that case and gets confused).

Reply:
In that case we should add the appropriate typesheds to the pre-commit extras.

Reply:
Example from another project: https://github.com/glotzerlab/signac/blob/3d546da97da810b55ce7d03028b692d8dd743b23/.pre-commit-config.yaml#L59

Reply:
The issue (is it an issue?) here is that numpy typestubs are now _in_ numpy (so the only way to get them is to install numpy in the pre-commit venv). If that is reasonable, this is the additional diff I need. It's mostly working around the issue that untagged union types are arguably an anti-pattern, but python has no support for tagged unions and so this is where we are

```patch
diff --git a/python/cudf/cudf/core/column/column.py b/python/cudf/cudf/core/column/column.py
index 194377a7c9..0b963c5506 100644
--- a/python/cudf/cudf/core/column/column.py
+++ b/python/cudf/cudf/core/column/column.py
@@ -1057,7 +1057,7 @@ class ColumnBase(Column, Serializable, BinaryOperand, Reducible):
         frames = []
         header["type-serialized"] = pickle.dumps(type(self))
         try:
-            dtype, dtype_frames = self.dtype.serialize()
+            dtype, dtype_frames = self.dtype.serialize()  # type: ignore[union-attr]
             header["dtype"] = dtype
             frames.extend(dtype_frames)
             header["dtype-is-cudf-serialized"] = True
diff --git a/python/cudf/cudf/core/column/numerical.py b/python/cudf/cudf/core/column/numerical.py
index 4b74dde129..7fc4e8b81a 100644
--- a/python/cudf/cudf/core/column/numerical.py
+++ b/python/cudf/cudf/core/column/numerical.py
@@ -591,11 +591,13 @@ class NumericalColumn(NumericalBaseColumn):
             else:
                 # Kinds are the same but to_dtype is smaller
                 if "float" in to_dtype.name:
-                    finfo = np.finfo(to_dtype)
+                    finfo: Any = np.finfo(to_dtype)
                     lower_, upper_ = finfo.min, finfo.max
                 elif "int" in to_dtype.name:
-                    iinfo = np.iinfo(to_dtype)
+                    iinfo: Any = np.iinfo(to_dtype)
                     lower_, upper_ = iinfo.min, iinfo.max
+                else:
+                    raise RuntimeError("Unhandled kind")
 
                 if self.dtype.kind == "f":
                     # Exclude 'np.inf', '-np.inf'
@@ -634,7 +636,7 @@ class NumericalColumn(NumericalBaseColumn):
 
         # want to cast int to float
         elif self.dtype.kind in {"i", "u"} and to_dtype.kind == "f":
-            info = np.finfo(to_dtype)
+            info: Any = np.finfo(to_dtype)
             biggest_exact_int = 2 ** (info.nmant + 1)
             if (self.min() >= -biggest_exact_int) and (
                 self.max() <= biggest_exact_int
diff --git a/python/cudf/cudf/core/column/timedelta.py b/python/cudf/cudf/core/column/timedelta.py
index e6d688014f..c496482d38 100644
--- a/python/cudf/cudf/core/column/timedelta.py
+++ b/python/cudf/cudf/core/column/timedelta.py
@@ -3,7 +3,7 @@
 from __future__ import annotations
 
 import datetime
-from typing import Any, Sequence, cast
+from typing import Any, Sequence, Union, cast
 
 import numpy as np
 import pandas as pd
@@ -164,7 +164,7 @@ class TimeDeltaColumn(ColumnBase):
             return NotImplemented
 
         this: ColumnBinaryOperand = self
-        out_dtype = None
+        out_dtype: Any = None
 
         if is_timedelta64_dtype(other.dtype):
             # TODO: pandas will allow these operators to work but return false
@@ -224,6 +224,7 @@ class TimeDeltaColumn(ColumnBase):
             if np.isnat(other):
                 return cudf.Scalar(None, dtype=self.dtype)
 
+            common_dtype: Union[str, Dtype]
             if other_time_unit not in {"s", "ms", "ns", "us"}:
                 common_dtype = "timedelta64[s]"
             else:
diff --git a/python/cudf/cudf/core/join/_join_helpers.py b/python/cudf/cudf/core/join/_join_helpers.py
index 790f295199..2589c29fad 100644
--- a/python/cudf/cudf/core/join/_join_helpers.py
+++ b/python/cudf/cudf/core/join/_join_helpers.py
@@ -72,9 +72,9 @@ def _match_join_keys(
         if left_is_categorical:
             if how in {"left", "leftsemi", "leftanti"}:
                 return lcol, rcol.astype(ltype)
-            common_type = ltype.categories.dtype
+            common_type = cast(CategoricalDtype, ltype).categories.dtype
         else:
-            common_type = rtype.categories.dtype
+            common_type = cast(CategoricalDtype, rtype).categories.dtype
         return lcol.astype(common_type), rcol.astype(common_type)
 
     if is_dtype_equal(ltype, rtype):
```

I'm -epsilon on adding numpy to the mypy environment, because it's not typed well enough, and I think will create more friction than is necessary.

Reply:
Hmm. This could go either way -- my preference would be to have the results of `pre-commit run mypy [--all-files]` align with the results of running `mypy` directly. However, the introduction of a NumPy dependency in the pre-commit environment is less than ideal. I defer to your judgement on this, or perhaps @vyasr if you'd like to weigh in.

Reply:
I think the problem with that is where do you stop? pandas also maintain type annotations and type-stubs within pandas, rather than spun out, so do we also need pandas in the pre-commit environment, and then...

I don't think there's a nice way to solve this. For something like `clang-format` via `pre-commit` all we ask for is the binary, and rely on the dev environment to provide the headers. In contrast, installing a sandboxed `mypy` needs us to inject the conda dev environment's `sys.path` into `MYPYPATH` to get consistency between `mypy` and `pre-commit run mypy`.

Reply:
> where do you stop?

You would install all packages from [this list](https://github.com/rapidsai/cudf/blob/branch-22.10/conda/environments/cudf_dev_cuda11.5.yml) that have typing stubs that we care about validating against. To the best of my knowledge, that list would likely include numpy and pandas -- I'd bet we would be happy with only those two packages.

I don't quite follow the discussion of `clang-format` and injecting a conda environment path for mypy. We certainly don't want to mix a user's conda environment with the pre-commit managed environments. My intent was to use `additional_dependencies` to specify a small number of additional packages that should be used solely for typing stubs such that `mypy` in a user conda environment built from the yml linked above would act similarly to `pre-commit run mypy`.

I am happy to defer this conversation for later, since it can be handled separately from the main goal of updating to mypy 0.971.

Reply:
> I don't quite follow the discussion of `clang-format` and injecting a conda environment path for mypy. 

Ah, sorry, I was thinking `clangd` editor type-checking. Of course, there's no need to have linting based type-checking for the C++ code (since if it compiles it type-checks). I think what I meant was that if the developer configures their editor to use live type hints they will probably not do so using the `pre-commit` provided mypy (because you can't figure out the path to it programmatically), so they'll just use a version that's installed in the development environment which may pick up other packages.

> I don't know of cudf relying on typing stubs from cupy or dask or anything else.

While cudf may not _rely_ on them, the problem appears as soon as the module is importable and provides types to mypy. Where previously a function would have deduced return type `Any` (which always provides any attributes), it will now have a narrower type which might mean type-checking fails.

> I am happy to defer this conversation for later, since it can be handled separately from the main goal of updating to mypy 0.971.

Yeah, let's do that separately (https://github.com/rapidsai/cudf/issues/11661)

Reply:
I'm late to the game here, and I'll follow up in more depth on #11661, but I agree with @wence- that this is trickier than it seems. `cupy` in particular would become a problem, since we've definitely had e.g. docs-only contributions before where users didn't even need to have CUDA or cudf-compatible GPU on their machine.